File: flinkml/src/main/scala/de/tuberlin/dima/bdapro/flinkml/recommendation/ALSRating.scala
package de.tuberlin.dima.bdapro.flinkml.recommendation

import breeze.linalg.Axis._1
import de.tuberlin.dima.bdapro.flinkml.Config
import org.apache.flink.api.scala.{DataSet, _}
import org.apache.flink.ml.common.{LabeledVector, ParameterMap}
import org.apache.flink.ml.math.Vector
import org.apache.flink.ml.recommendation.ALS
import org.apache.flink.ml.preprocessing.Splitter

/**
 * Class for testing the ALS Rating (Recommendation) ML algorithm
 * 
 * @author Seema Narasimha Swamy
 *
 */
class ALSRating (val envPassed : ExecutionEnvironment) {
  val env = envPassed

  /**
   * 
   * The execute method is used to test the algorithm.
   * The input data set (movielens) is in CSV format which is split into two parts 80% for learning and the rest for testing.
   * The method returns "Root Mean Square Error" for the algorithm.
   * 
   */
  def execute(): Double = {

    val pathToDataset = Config.pathToRecommendationTrainingSet
    
      /**
       * Read input data set from a csv file
       */

      val inputDS: DataSet[(Int, Int, Double)] = env
        .readCsvFile[(Int, Int, Double)](pathToDataset, ignoreFirstLine = true)

      /*
       * Setup the ALS learner.
       */
      val als = ALS()
        .setIterations(10)
        .setNumFactors(10)
        .setBlocks(100)

      /*
       * Set the other parameters via a parameter map
       */
      val parameters = ParameterMap()
        .add(ALS.Lambda, 0.9)
        .add(ALS.Seed, 42L)
  
      /*
  		 * Split the data into training and test sets (80% training and 20% held for testing).
  		 */
      val trainTestData = Splitter.trainTestSplit(inputDS, 0.8, true)
      val trainingData: DataSet[(Int, Int, Double)] = trainTestData.training
      val testingData: DataSet[(Int, Int, Double)] = trainTestData.testing
      
      /*
       * Fit the model.
       */
      als.fit(trainingData, parameters)

      // ********* For Testing the model *************** //

      /*
       * Calculate the ratings according to the matrix factorization
       * Evaluate the algorithm and find the test error.
       */
      val predictedRatings = als.predict(testingData.map(x => (x._1, x._2)))
      
      var foo = 0.9
      val predictionsAndRatings: DataSet[(Double, Double)] =
          testingData.join(predictedRatings).where(0, 1).equalTo(0, 1) { (l, r) => (l._3, r._3) }
  
      val squaredError: DataSet[Double] = predictionsAndRatings.map((tuple: (Double, Double)) => ((tuple._1 - tuple._2) * (tuple._1 - tuple._2), 1))
        .reduce((tuple: (Double, Int), tuple0: (Double, Int)) => (tuple._1 + tuple0._1, tuple._2 + tuple0._2))
        .map(tuple => (math.sqrt(tuple._1 / tuple._2)))
      val result = squaredError.collect()
      
      return result.last

  }
}

File: sparkml/src/main/java/de/tuberlin/dima/bdapro/sparkml/recommendation/ALSRating.java
package de.tuberlin.dima.bdapro.sparkml.recommendation;

import de.tuberlin.dima.bdapro.sparkml.Config;
import de.tuberlin.dima.bdapro.sparkml.MLAlgorithmBase;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import java.io.Serializable;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.ml.evaluation.RegressionEvaluator;
import org.apache.spark.ml.recommendation.ALS;
import org.apache.spark.ml.recommendation.ALSModel;

/**
 * Class for testing the ALS Rating (Recommendation) ML algorithm
 * 
 * @author Seema Narasimha Swamy
 *
 */
public class ALSRating extends MLAlgorithmBase {
	public ALSRating(final SparkSession spark) {
		super(spark);
	}

	public static class Rating implements Serializable {
		private static final long serialVersionUID = 1L;
		
		private int userId;
		private int movieId;
		private float rating;
		private long timestamp;

		public Rating() {
		}

		public Rating(int userId, int movieId, float rating, long timestamp) {
			this.userId = userId;
			this.movieId = movieId;
			this.rating = rating;
			this.timestamp = timestamp;
		}

		public int getUserId() {
			return userId;
		}

		public int getMovieId() {
			return movieId;
		}

		public float getRating() {
			return rating;
		}

		public long getTimestamp() {
			return timestamp;
		}

		public static Rating parseRating(String str) {
			String[] fields = str.split(",");
			if (fields.length != 4) {
				throw new IllegalArgumentException("Each line must contain 4 fields");
			}
			int userId = Integer.parseInt(fields[0]);
			int movieId = Integer.parseInt(fields[1]);
			float rating = Float.parseFloat(fields[2]);
			long timestamp = Long.parseLong(fields[3]);
			return new Rating(userId, movieId, rating, timestamp);
		}
	}

	
	/**
     * 
     * The execute method is used to test the algorithm.
     * The input data set (movielens) is in CSV format which is split into two parts 80% for learning and the rest for testing.
     * The method returns "Root Mean Square Error" for the algorithm.
     * 
     */
	public double execute() {
		/**
		 * Read input data set from a csv file
		 */
		String inputfile = Config.pathToRecommendationTrainingSet();
		JavaRDD<Rating> ratingsRDD = spark.read().textFile(inputfile).javaRDD().map(Rating::parseRating).cache();
		Dataset<Row> ratings = spark.createDataFrame(ratingsRDD, Rating.class);
		
		/*
		 * Split the data into training and test sets (80% training and 20% held for testing).
		 */
		Dataset<Row>[] splits = ratings.randomSplit(new double[] { 0.8, 0.2 });
		Dataset<Row> training = splits[0];
		Dataset<Row> test = splits[1];

		/*
		 *  Build the recommendation model using ALS on the training data
		 */
		ALS als = new ALS().setMaxIter(5).setRegParam(0.01).setUserCol("userId").setItemCol("movieId")
				.setRatingCol("rating");
		ALSModel model = als.fit(training);

		/*
		 * Evaluate the model by computing the RMSE on the test data
		 * Note we set cold start strategy to 'drop' to ensure we don't get NaN
		 * evaluation metrics
		 * 
		 */ 
		model.setColdStartStrategy("drop");
		
		/*
         * Make predictions.
         */
		Dataset<Row> predictions = model.transform(test);

		/*
         *  Evaluate prediction and compute Root Mean Square Error.
         */
		RegressionEvaluator evaluator = new RegressionEvaluator().setMetricName("rmse").setLabelCol("rating")
				.setPredictionCol("prediction");
		Double rmse = evaluator.evaluate(predictions);
		
		return rmse;
	}
}
